src/bin/pg_autoctl/supervisor.c

```diff
@@ -0,0 +1,880 @@
+	{
+		pid_t pid;
+		int status;
+
+		/* Check that we still own our PID file, or quit now */
+		(void) check_pidfile(supervisor->pidfile, supervisor->pid);
+
+		/* If necessary, now is a good time to reload services */
+		if (asked_to_reload)
+		{
+			log_info("pg_autoctl received a SIGHUP signal, "
+					 "reloading configuration");
+			(void) supervisor_reload_services(supervisor);
+		}
+
+		if (firstLoop)
+		{
+			firstLoop = false;
+		}
+		else
+		{
+			/* avoid buzy looping on waitpid(WNOHANG) */
+			pg_usleep(100 * 1000); /* 100 ms */
+		}
+
+		/* ignore errors */
+		pid = waitpid(-1, &status, WNOHANG);
+
+		switch (pid)
+		{
+			case -1:
+			{
+				if (errno == ECHILD)
+				{
+					/* no more childrens */
+					if (asked_to_stop || asked_to_stop_fast)
+					{
+						/* off we go */
+						log_info("Internal subprocesses are done, stopping");
+						return true;
+					}
+				}
+				else
+				{
+					log_fatal("Failed to call waitpid(): %m");
+					return false;
+				}
+			}
+
+			case 0:
+			{
+				/*
+				 * We're using WNOHANG, 0 means there are no stopped or exited
+				 * children, it's all good. It's the expected case when
+				 * everything is running smoothly, so enjoy and sleep for
+				 * awhile.
+				 *
+				 * That said, we could have received a signal that instructs a
+				 * shutdown, such as SIGTERM or SIGINT. In that case we need to
+				 * do one of these things:
+				 *
+				 * - first time we receive the signal, begin a shutdown
+				 *   sequence for all services and the main supervisor itself,
+				 *
+				 * - when receiving the signal again, if it's a SIGTERM
+				 *   continue the shutdown sequence,
+				 *
+				 * - when receiving a SIGINT forward it to our services so as
+				 *   to finish as fast as we can.
+				 *
+				 * Sending SIGTERM and then later SIGINT if the process is
+				 * still running is a classic way to handle service shutdown.
+				 */
+				bool shutdownNow = false;
+
+				if (asked_to_stop || asked_to_stop_fast)
+				{
+					if (!supervisor->shutdownSequenceInProgress)
+					{
+						log_info("pg_autoctl received signal %s, terminating",
+								 asked_to_stop_fast ? "SIGINT" : "SIGTERM");
+
+						supervisor->cleanExit = true;
+						supervisor->shutdownSequenceInProgress = true;
+
+						shutdownNow = true;
+					}
+					else if (asked_to_stop_fast)
+					{
+						/*
+						 * Even when we are already shutting down, we process
+						 * SIGINT again and log about it.
+						 */
+						log_info(
+							"pg_autoctl received signal SIGINT, terminating");
+
+						shutdownNow = true;
+					}
+				}
+
+				/*
+				 * Initiate the stop sequence the first time we receive SIGTERM
+				 * and then each time we receive SIGINT.
+				 */
+				if (shutdownNow)
+				{
+					/*
+					 * Stop all the services.
+					 */
+					if (stoppingLoopCounter == 0)
+					{
+						(void) supervisor_stop_subprocesses(supervisor);
+					}
+
+					(void) supervisor_shutdown_sequence(stoppingLoopCounter++);
```

Comment:
This should be in a separate if, because `shutdownNow` is not always true anymore now when we are shutting down. We need either a separate boolean, e.g. `shuttingDown` or we reuse stoppingLoopCounter like `stoppingLoopCounter > 0 || shutdownNow`.

Reply:
Done now.